americangut/diversity_analysis.py

```diff
@@ -0,0 +1,1729 @@
+    if int(max_hi*fudge) < 5:
+        lowest = np.ceil(max_hi*fudge/factor)/(fudge)*factor
+    else:
+        lowest = np.ceil(max_hi*fudge*10/factor)/(fudge*10)*factor
+
+    return lowest, fudge
+
+
+def _get_p_value(sub_p, ref_group, group, p_tab_col):
+    """Determines the comparison value within a post-hoc table"""
+    if '%s vs. %s' % (ref_group, group) in sub_p['Group 2'].values:
+        p_value = sub_p.loc[sub_p['Group 2'] == '%s vs. %s'
+                            % (ref_group, group), p_tab_col].values[0]
+    elif '%s vs. %s' % (ref_group, group) in sub_p['Group 1'].values:
+        p_value = sub_p.loc[sub_p['Group 1'] == '%s vs. %s'
+                            % (ref_group, group), p_tab_col].values[0]
+    elif '%s vs. %s' % (group, ref_group) in sub_p['Group 2'].values:
+        p_value = sub_p.loc[sub_p['Group 2'] == '%s vs. %s'
+                            % (group, ref_group), p_tab_col].values[0]
+    elif '%s vs. %s' % (group, ref_group) in sub_p['Group 1'].values:
+        p_value = sub_p.loc[sub_p['Group 1'] == '%s vs. %s'
+                            % (group, ref_group), p_tab_col].values[0]
+    else:
+        raise ValueError('%s vs. %s is not a defined group'
+                         % (ref_group, group))
+
+    return p_value
+
+
+def _correct_p_value(tail, p_value, ref_val, current_val):
+    """Determines if p-values should be tail corrected"""
+    if tail and ref_val > current_val:
+        return 1
+    else:
+        return p_value
+
+
+def get_distance_vectors(dm, df, group, order=None):
+    """Extracts the distance information for all samples in a group
+
+    Parameters
+    ----------
+    dm : skbio DistanceMatrix
+        A distance matrix object with the samples corresponding to those in
+        the data frame with the metadata.
+    df : pandas DataFrame
+        A dataframe containing the metadata associated withed with the object
+    group : str
+        the metadata category being interograted, used to group the data
+    order : None, list
+        Default is None. The order of groups in group.
+
+    Returns
+    -------
+    within : list
+        A list of the group names for the within group distance
+    w_dist : list
+        A list of arrays with the within group distances
+    between : list
+        A list of the group names for between group distance
+    b_dist : list
+        A list of arrays for between group distance
+
+    """
+    # Checks the column is supported
+    if group not in df:
+        raise ValueError('%s is not a metadata category.' % group)
+
+    # Gets the sample ids associated with the group
+    if order is None:
+        order = list(df.groupby(group).groups)
+    ordered_ids = {o: df.groupby(group).groups[o] for o in order}
+
+    # Gets the data
+    dist_ids = np.array(list(dm.ids))
+    dist_data = dm.data
+
+    # Prealocates objects for return
+    within = []
+    w_dist = []
+    between = []
+    b_dist = []
```

Comment:
It would take a bit of thought, but it should be possible to determine the size of these objects. What that means is you could preallocate (e.g., `b_dist = zeros(the_number_of_resulting_elements, dtype=float)`) and then just fill the vectors. It would rework this some, but if this code is slow (suggest profiling first), then it would be one way to definitely speed it up. The `append` calls in the inner loop, if done a lot, will get expensive particularly if the resulting `list`s are large in size due to how Python allocates memory